src/pyhf/contrib/viz/brazil.py

```diff
@@ -18,33 +279,115 @@ def plot_results(ax, mutests, tests, test_size=0.05):
         >>> data = observations + model.config.auxdata
-        >>> poi_vals = np.linspace(0, 5, 41)
+        >>> test_pois = np.linspace(0, 5, 41)
         >>> results = [
         ...     pyhf.infer.hypotest(test_poi, data, model, return_expected_set=True)
-        ...     for test_poi in poi_vals
+        ...     for test_poi in test_pois
+        ... ]
+        >>> fig, ax = plt.subplots()
+        >>> artists = pyhf.contrib.viz.brazil.plot_results(test_pois, results, ax=ax)
+
+        A Brazil band plot with the components of the :math:`\mathrm{CL}_{s}`
+        ratio drawn on top.
+
+        >>> import numpy as np
+        >>> import matplotlib.pyplot as plt
+        >>> import pyhf
+        >>> import pyhf.contrib.viz.brazil
+        >>> pyhf.set_backend("numpy")
+        >>> model = pyhf.simplemodels.hepdata_like(
+        ...     signal_data=[12.0, 11.0], bkg_data=[50.0, 52.0], bkg_uncerts=[3.0, 7.0]
+        ... )
+        >>> observations = [51, 48]
+        >>> data = observations + model.config.auxdata
+        >>> test_pois = np.linspace(0, 5, 41)
+        >>> results = [
+        ...     pyhf.infer.hypotest(
+        ...         test_poi, data, model, return_expected_set=True, return_tail_probs=True
+        ...     )
+        ...     for test_poi in test_pois
         ... ]
         >>> fig, ax = plt.subplots()
-        >>> pyhf.contrib.viz.brazil.plot_results(ax, poi_vals, results)
+        >>> artists = pyhf.contrib.viz.brazil.plot_results(
+        ...     test_pois, results, ax=ax, components=True
+        ... )
 
     Args:
-        ax (`matplotlib.axes.Axes`): The matplotlib axis object to plot on.
-        mutests (:obj:`list` or :obj:`array`): The values of the POI where the
+        test_pois (:obj:`list` or :obj:`array`): The values of the POI where the
           hypothesis tests were performed.
-        tests (:obj:`list` or :obj:`array`): The :math:`\\mathrm{CL}_{s}` values
-          from the hypothesis tests.
-        test_size (:obj:`float`): The size, :math:`\\alpha`, of the test.
+        tests (:obj:`list` or :obj:`array`): The collection of :math:`p`-value-like
+          values (:math:`\mathrm{CL}_{s}` values or tail probabilities) from
+          the hypothesis tests.
+          If the ``components`` keyword argument is ``True``,  ``tests`` is required
+          to have the same structure as :func:`pyhf.infer.hypotest`'s return
+          when using ``return_expected_set=True`` and ``return_tail_probs=True``:
+          a tuple of :math:`\mathrm{CL}_{s}`,
+          :math:`\left[\mathrm{CL}_{s+b}, \mathrm{CL}_{b}\right]`,
+          :math:`\mathrm{CL}_{s,\mathrm{exp}}` band.
+        test_size (:obj:`float`): The size, :math:`\alpha`, of the test.
+        ax (:obj:`matplotlib.axes.Axes`): The matplotlib axis object to plot on.
+
+    Returns:
+        :class:`BrazilBandArtist`: Artist containing the :obj:`matplotlib.artist`
+        objects drawn.
     """
+    if ax is None:
+        ax = plt.gca()
```

Comment:
given that at this point we *know* for sure what axes all the artists will be attached on I feel we could skip the logic in the `BrazilBandCollection` and just pass the axes as an ctor argument instead of recovering it from the artists

Reply:
Hm, let me think on this a bit. 

> given that at this point we know for sure what axes all the artists will be attached on

This is true inside the plotting functions, but as the `BrazilBandCollection` is in the public API there's nothing keeping someone from trying to do something confusing and wrong like we test against now in `test_brazil_band_collection`.

> just pass the axes as an ctor argument instead of recovering it from the artists

This would get rid of the problem of assigning artists from multiple axes to the same `BrazilBandCollection`, but then you now have a namedtuple that contains both the artists as well as the `axes.Axes` object they're on.

Though at the same time, that API for all the plotting functions basically are expecting this sort of info so maybe this is pretty reasonable.

Reply:
Okay, convinced. :)